lib/Sema/CSDiag.cpp

```diff
@@ -2153,6 +2153,8 @@ static void eraseOpenedExistentials(Expr *&expr, ConstraintSystem &CS) {
         return type;
       });
       CS.setType(expr, type);
+      // Set new type to the expression directly.
+      expr->setType(type);
```

Comment:
It's non-obvious why this is the right thing to do. This seems like an indication that something that should be reading from the type cache isn't.